rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+  (Vector core team) to run experiments and tests against locally. This doesn't
+  mean we want to automate this and include running tests locally against our
+  whole k8s test infrastructure - but the ability to do it with little effort is
+  very important: even if we employ super-reliable CI automation, the turnaround
+  time of going through it is way higher than conducting an experiment locally.
+  Locally means using local code tree and binaries - the infrastructure itself
+  is still in the cloud.
+
+- Ideally, we want the test system to be available not just to the Vector core
+  team, but to the whole open-source community. Of course, we don't want to give
+  unrestricted access to _our_ cloud testing infrastructure - but the solution
+  we employ should allow third-parties to bring their own resources. Things that
+  are local in essence (like `minikube`) should just work. There shouldn't be a
+  situation where one can't run tests in `minikube` because cloud parts aren't
+  available. We already have a similar constraints at the Vector Test Harness.
+
+- We need the required efforts to managements the solution to be low, and the
+  price to be relatively small. This means that the solution has to be simple.
+
+- We want to expose the same kind of interface to each of the clusters, so the
+  cluster we run the tests is easily interchangeable.
+  A kubectl config file is a good option, since it encapsulates all the
+  necessary information tp connect to a cluster.
+
+Based on all of the above, it makes sense to split the infrastructure into two
+parts.
+
+- Cloud infrastructure that we manage and pay for.
+
+  We will create a dedicated public repo with [Terraform] configs to setup a
+  long-running Kubernetes test infrastructure.
+  The goal here is to make the real, live cloud environments available for
+  people and automation to work with.
+
+- Self-hosted infrastructure that we maintain configs for.
+
+  This is what keep so that it's easy to run the a self-hosted cluster. Most
+  likely locally - for things like `minikube`, but not limited to. The focus
+  here is lock particular _versions_ and _configuration_ of the tooling, so it's
+  easy to run tests against. Potentially even having multiple versions of the
+  same tool, for instance, when you need to compare `minikube` `1.9.2` and
+  `1.8.2`.
+  The goal here is to address the problem of configuring the self hosted cluster
+  management tools once and for all, and share those configurations. For people
+  it has the benefit of enabling them to spend time on soling the problem (or
+  doing whatever they need to do with k8s) rather than spending time on
+  configuration. For automation flows - it'll make it really simple to reference
+  a particular self-hosted configuration - and offload the complexity of
+  preparing it.
+
+  This one we'll have to figure out, but most likely we'll create a dedicated
+  repo per tool, each with different rules - but with a single interface.
```

Comment:
Why wouldn't we just add these to the `docker-compose.yml` file, just like we do for other integration tests? I think it's very important we keep the workflow same for local integration test. Whether it's AWS S3, Kubernetes, Docker, Loki, we should be able to test them all locally with the proper service.